spec.html

```diff
@@ -22546,21 +22571,95 @@ <h1>ExecuteModule ( ) Concrete Method</h1>
+        <h1>Runtime Semantics: HostImportModuleDynamically ( _referencingScriptOrModule_, _specifier_, _promiseCapability_ )</h1>
+
+        <p>HostImportModuleDynamically is an implementation defined abstract operation that performs any necessary setup work in order to make available the module corresponding to the |ModuleSpecifier| String, _specifier_, occurring within the context of the script or module represented by the Script Record or Module Record _referencingScriptOrModule_. (_referencingScriptOrModule_ may also be *null*, if there is no active script or module when the `import()` expression occurs.) It then performs FinishDynamicImport to finish the dynamic import process.</p>
+
+        <p>The implementation of HostImportModuleDynamically must conform to the following requirements:</p>
+
+        <ul>
+          <li>
+            The abstract operation must always complete normally with *undefined*. Success or failure must instead be signaled as discussed below.
+          </li>
+          <li>
+            The host environment must conform to one of the two following sets of requirements:
+            <dl>
+              <dt>Success path</dt>
+
+              <dd>
+                <ul>
+                  <li>At some future time, the host environment must perform FinishDynamicImport(_referencingScriptOrModule_, _specifier_, _promiseCapability_, NormalCompletion(*undefined*)).</li>
+
+                  <li>Any subsequent call to HostResolveImportedModule after FinishDynamicImport has completed, given the arguments _referencingScriptOrModule_ and _specifier_, must complete normally.</li>
+
+                  <li>The completion value of any subsequent call to HostResolveImportedModule after FinishDynamicImport has completed, given the arguments _referencingScriptOrModule_ and _specifier_, must be a module which has already been evaluated, i.e. whose Evaluate concrete method has already been called and returned a normal completion.</li>
+                </ul>
+              </dd>
+
+              <dt>Failure path</dt>
+
+              <dd>
+                <ul>
+                  <li>At some future time, the host environment must perform FinishDynamicImport(_referencingScriptOrModule_, _specifier_, _promiseCapability_, an abrupt completion), with the abrupt completion representing the cause of failure.</li>
+                </ul>
+              </dd>
+            </dl>
+
+            <p>The intent of this specification is to not violate run to completion semantics. The spec-level formalization of this is a work-in-progress.</p>
+          </li>
+          <li>
+            Every call to HostImportModuleDynamically with the same _referencingScriptOrModule_ and _specifier_ arguments must conform to the <em>same</em> set of requirements above as previous calls do. That is, if the host environment takes the success path once for a given _referencingScriptOrModule_, _specifier_ pair, it must always do so, and the same for the failure path.
```

Comment:
Does this also hold with a null referencingScriptOrModule? ie, is everything with `null` considered the same referencing module, or is it more like NaN where they’re all intended to be distinct?

Reply:
Yes, it does hold. Do you think this is unclear in the text?

Reply:
My reading is that for a given referencing location, two calls should be idempotent, but that one call from each of two different locations need not produce the same result.

This line suggests that one call from each of two different locations with null will always be required to have the same result - ie, it seems *stricter* on engines for when there’s no referencing location. The difference between the cases strikes me as odd; it seems like either the null case needs relaxing or the non-null case needs tightening, to make things more consistent?

Reply:
Well, to explain how it works out in HTML (cc @domenic), if you are in one of these null contexts, then it's basically like you're in the context of the whole web page. So the base URL of the whole page is used to parse relative URLs. In this case, it *does* make sense to group together all the null imports.

I suspect it might work out similarly in other environments: When you don't have a clear referencing script, the context is something corresponding to the Realm. But it may be that there are other environments that need a weaker condition.

How about we start with this strong condition, and consider a follow-on needs-consensus PR to weaken the condition if an embedding environment comes along where this invariant needs to be relaxed? Stating as strong conditions as possible could keep things more predictable for ECMAScript itself.

Reply:
Sounds good - if the intended usage in browsers is indeed to group all nulls as the same effective referencing script, then as-written seems fine to me.